device.js

```diff
@@ -139,36 +144,46 @@ DeviceTree.prototype.getDirectory = function (qnode) {
     return new Promise((resolve, reject) => {
         self.addRequest((error) => {
             if (error) {
-                reject(error);
                 self.finishRequest();
+                reject(error);
                 return;
             }
 
-            let cb = (error, node) => {
-                self.clearTimeout(); // clear the timeout now. The resolve below may take a while.
+            self.callback = (error, node) => {
+                const requestedPath = qnode.path != null ? qnode.path : qnode.elements["0"].path;
                 if (error) {
+                    if (self._debug) {
+                        console.log("Received getDirectory error", error);
+                    }
+
+                    self.clearTimeout(); // clear the timeout now. The resolve below may take a while.
```

Comment:
I decided to move `self.finishRequest()` after `resole`/`reject` to ensure that in if in the meanwhile a client was re-requesting a `getDirectory` it would not erase the newly created callback.